drivers/i2c/i2c_gd32.c

```diff
@@ -0,0 +1,684 @@
+		 */
+		if ((current->flags & I2C_MSG_RW_MASK) !=
+		    (next->flags & I2C_MSG_RW_MASK)) {
+			if ((next->flags & I2C_MSG_RESTART) == 0U) {
+				return -EINVAL;
+			}
+		}
+
+		/* Only the last message need I2C_MSG_STOP flag to free the Bus. */
+		if (current->flags & I2C_MSG_STOP) {
+			return -EINVAL;
+		}
+
+		current++;
+	}
+
+	/* Last message flags implict contain I2C_MSG_STOP flag. */
+	current->flags |= I2C_MSG_STOP;
+
+	k_sem_take(&data->bus_mutex, K_FOREVER);
+
+	if (data->dev_config & I2C_ADDR_10_BITS) {
+		data->addr1 = 0xF0 | ((addr & BITS(8, 9)) >> 8U);
+		data->addr2 = addr & BITS(0, 7);
+	} else {
+		data->addr1 = addr & BITS(0, 6);
+	}
+
+	for (uint8_t i = 0; i < num_msgs; i = itr) {
+		data->current = &msgs[i];
+		data->xfer_len = msgs[i].len;
+
+		for (itr = i + 1; itr < num_msgs; itr++) {
+			if ((data->current->flags & I2C_MSG_RW_MASK) !=
+			    (msgs[itr].flags & I2C_MSG_RW_MASK)) {
+				break;
+			}
+			data->xfer_len += msgs[itr].len;
+		}
+
+		if (data->current->flags & I2C_MSG_READ) {
+			err = i2c_gd32_msg_read(dev);
+		} else {
+			err = i2c_gd32_msg_write(dev);
+		}
+
+		if (err < 0) {
+			break;
+		}
+	}
+
+	k_sem_give(&data->bus_mutex);
+
+	return err;
+}
+
+static int i2c_gd32_configure(const struct device *dev,
+			      uint32_t dev_config)
+{
+	struct i2c_gd32_data *data = dev->data;
+	const struct i2c_gd32_config *cfg = dev->config;
+	uint32_t pclk1, freq, clkc;
+	int err = 0;
+
+	k_sem_take(&data->bus_mutex, K_FOREVER);
+
+	i2c_gd32_disable(cfg);
+
+	/* GD32 i2c interface always connect to APB1. */
+	pclk1 = rcu_clock_freq_get(CK_APB1);
+
+	/* i2c clock frequency, us */
+	freq = pclk1 / 1000000U;
+	if (freq > I2CCLK_MAX) {
+		LOG_ERR("I2C max clock freq %u, current is %u\n",
+			I2CCLK_MAX, freq);
+		err = -ENOTSUP;
+		goto error;
+	}
+
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		if (freq < I2CCLK_SM_MIN) {
+			LOG_ERR("I2C standard-mode min clock freq %u, current is %u\n",
+				I2CCLK_SM_MIN, freq);
+			err = -ENOTSUP;
+			goto error;
+		}
+		I2C_CTL1(cfg->reg) &= ~I2C_CTL1_I2CCLK;
+		I2C_CTL1(cfg->reg) |= freq;
+
+		/* Standard-mode risetime maximum value: 1000ns */
+		if (freq == I2CCLK_MAX) {
+			I2C_RT(cfg->reg) = I2CCLK_MAX;
+		} else {
+			I2C_RT(cfg->reg) = freq + 1U;
+		}
+
+		/* pclk1 = bitrate * (clkc * (t_high_scale_1 + t_low_scale_1)) */
+		clkc = pclk1 / (I2C_BITRATE_STANDARD * 2U);
+
+		/* clear and set clkc */
+		I2C_CKCFG(cfg->reg) &= ~I2C_CKCFG_CLKC;
+		I2C_CKCFG(cfg->reg) |= clkc;
+		/* standard-mode */
+		I2C_CKCFG(cfg->reg) &= ~I2C_CKCFG_FAST;
+
+		break;
+	case I2C_SPEED_FAST:
+		if (freq < I2CCLK_FM_MIN) {
+			LOG_ERR("I2C fast-mode min clock freq %u, current is %u\n",
+				I2CCLK_FM_MIN, freq);
+			err = -ENOTSUP;
+			goto error;
+		}
+
+		/* Fast-mode risetime maximum value: 300ns */
+		I2C_RT(cfg->reg) = freq * 300U / 1000U + 1U;
+
+		/* pclk1 = bitrate * (clkc * (t_high_scale_9 + t_low_scale_16)) */
+		clkc = pclk1 / (I2C_BITRATE_FAST_PLUS * 25);
```

Comment:
nit: some magic constants here, better define them (with explanation)

Reply:
/*
	 * From GD32 SoC user manual I2C_CKCFG register description.
	 * For standard mode:
	 *   T_high = CLKC * T_pclk1
	 *   T_low  = CLKC * T_pclk1
	 *   T_pclk1 = 1 / pclk1;
	 * 
	 * For fast mode and fast mode plus:
	 *   T_high = 9 * CLKC * T_pclk1
	 *   T_low  = 16 * CLKC * T_pclk1
	 *   T_pclk1 = 1 / pclk1;
	 *
	 * For a signal bit transfer: 
	 *  T_high + T_low = 1 / bitrate
	 * 
	 * CLKC equation for standard mod:
	 *   CLKC = pclk1 / (bitrate * 2)
	 * 
	 * CLKC equation for fast mode and fast mode plus:
	 *   CLKC = pclk1 / (bitrate * 25)
	 *
	 * T_high: high period of the SCL clock
	 * T_low: low period of the SCL clock
	 * T_pclk1: duration of signal pclk1 pulse
	 * pclk1: i2c device clock frequency
	 * bitrate: 100 Kbits for standard mode
	 */

Reply:
Is that explanation clear?

Reply:
done.